### PR TITLE
Allows JSON unserializing to be associative or object

### PR DESCRIPTION
**Pull request by jajajaime**
### Description (*)

Allows a developer to pass an optional boolean argument when unserializing a JSON string, so it can be returned as object or associative. Keeps associative as default for backwards compatibility.

### Manual testing scenarios (*)

1. Create code that unserializes json string while passing the extra argument of false

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
I remember we had some architecture discussions in https://github.com/magento/architecture/pull/254, https://github.com/magento/magento-coding-standard/pull/91 and 
 previously in https://github.com/magento/architecture/pull/161.

AFAIK no real decision was done yet. @lenaorobei do we have any updates on that?

File: lib/internal/Magento/Framework/Serialize/Serializer/Json.php [hunk 1]
```diff
@@ -32,9 +32,9 @@ public function serialize($data)
      * @inheritDoc
      * @since 100.2.0
      */
-    public function unserialize($string)
+    public function unserialize($string, $assoc = true)
```
Comment (orlangur) [inline lib/internal/Magento/Framework/Serialize/Serializer/Json.php:35]: Such change is not allowed: https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/

Let's first find out whether we need such functionality at all.

**Review by orlangur — CHANGES_REQUESTED** [review]

**Comment by ihor-sviziev** [comment]
@orlangur please explain why you moved this PR to "on hold" status.

**Comment by lenaorobei** [comment]
Unfortunately, we cannot accept this PR since it violates backward compatibility policy.

There were some discussions regarding introduction new JSON contract, but there is no resolution yet. Please see details here https://github.com/magento/architecture/pull/254

**Comment by jajajaime** [comment]
@lenaorobei is this because of "Adding parameters in public methods"? If so, I think it should be on a case by case basis, as creating a new class or interface would create more work for devs to switch to the new one in this case.

**Comment by orlangur** [comment]
@jajajaime backward compatibility policy is mandatory, no exception.

> when unserializing a JSON string, so it can be returned as object or associative

Please create a proposal in architecture repo first to describe why it needs to be returned as object.

**Pull request closed without merging**